### ISSUE TITLE
Ordering issue when cancelling consecutive requests

### ISSUE THREAD

**Issue by henryju** [microsoft/vscode-languageserver-node#522]
Hi,

I'm sorry if I'm reporting this issue at the wrong place. I'm using the vscode-language server client in our VSCode extension, and lsp4j on the server side.
When opening a file, the client send many requests in a row:

```
[Trace - 20:12:28] Sending notification 'textDocument/didOpen'.
Params: {
    "textDocument": {
        "uri": "file:///myfile",
[...]
```
```
[Trace - 20:12:28] Sending request 'textDocument/codeAction - (5)'.
Params: {
    "textDocument": {
        "uri": "file:///myfile"
    },
    "range": {
        "start": {
            "line": 0,
            "character": 0
        },
        "end": {
            "line": 0,
            "character": 0
        }
    },
    "context": {
        "diagnostics": []
    }
}
```
```
[Trace - 20:12:28] Sending request 'textDocument/codeAction - (6)'.
Params: {
    "textDocument": {
        "uri": "file:///myfile"
    },
    "range": {
        "start": {
            "line": 28,
            "character": 12
        },
        "end": {
            "line": 28,
            "character": 12
        }
    },
    "context": {
        "diagnostics": []
    }
}
```

```
[Trace - 20:12:28] Sending notification '$/cancelRequest'.
Params: {
    "id": 5
}
```

The fact two codeAction requests are sent when opening a file may be a bug in itself, but that's not my point here. The problem is the cancel request that is sent after the second codeAction. It increase the chances to have the first request already processed. Shouldn't they be swapped?

FYI here are the next logs:

```
[Trace - 20:12:29] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///myfile",
    "diagnostics": []
}
```

```
[Trace - 20:12:29] Received response 'textDocument/codeAction - (5)' in 716ms.
Result: []
```

```
[Trace - 20:12:29] Received response 'textDocument/codeAction - (6)' in 469ms.
Result: []
```

```
sept. 19, 2019 8:12:29 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 5
```

**Comment by dbaeumer** [comment]
@henryju I am not sure I understand correctly what the issue is. If the server has completed the result for (5) and nothing has changed for (6) the server is allowed to return a cached result. 

In the concrete example the first code action request is for position (0,0) whereas the second is for a different position (very likely the editor has move / set the cursor). So IMO all these requests do make sense.

**Comment by henryju** [comment]
My concern is about the cancel request of (5) that is sent after asking for (6). Shouldn't it be the other way around? When the editor ask for new codeAction position (6) it already knows that it is no more interested by the result of (5).

So basically, instead of:
(5)
(6)
cancel (5)

Why not:
(5)
cancel (5)
(6)

**Comment by dbaeumer** [comment]
That is fair but the current behavior is absolutely valid. The LSP stayed away from requiring things like this since they highly depend on how an editor works. Since the LSP library more or less forwards what the VS Code editor signals this would need to be addressed there. @henryju can you reproduce this as well using the VS Code API only.

**Comment by henryju** [comment]
> can you reproduce this as well using the VS Code API only
If you mean using VSCode API to listen to codeActions / cancel events, I haven't try (I was not even aware that the concept of cancelling an event was part of VSCode API).

**Comment by dbaeumer** [comment]
Yes, it is. The LSP itself doesn't cancel any events since it doesn't know if the client might want to have both results which is absolutely valid in terms of LSP. 

If you don't have any objection I will close the issue since IMO there is nothing we should change here.

**Comment by henryju** [comment]
Fine for me. Thanks for the clarification.
